docker/efficiency/builds.py

```diff
@@ -0,0 +1,221 @@
+import six
+import tarfile
+import collections
+
+from . import commons
+from .. import errors
+
+
+build_success_re = r'^Successfully built ([a-f0-9]+)\n$'
+
+# these prefixes are treated as remote by the docker daemon
+# (ref: pkg/urlutil/*) as of v1.6.0
+REMOTE_CONTEXT_PREFIXES = ["http://",
+                           "https://",
+                           "git://",
+                           "git@",
+                           "github.com/"]
+
+
+def get_build_id(build_result, discard_logs=False):
+    """ **Params:**
+        * `build_result` is a python generator returned by `Client.build`
+        * `discard_logs` (bool, default=False). If True, log lines will
+          be discarded after they're processed. Limits memory footprint.
+        **Returns** tuple:
+            1. Image ID if found, None otherwise
+            2. List of log lines
+    """
+    parsed_lines = []
+    image_id = None
+    for line in build_result:
+        try:
+            parsed_line = json.loads(line).get('stream', '')
+            if not discard_logs:
+                parsed_lines.append(parsed_line)
+            match = re.match(build_success_re, line)
+            if match:
+                image_id = match.group(1)
+        except ValueError:
+            # sometimes all the data is sent on a single line
+            # This ONLY works because every line is formatted as
+            # {"stream": STRING}
+            lines = re.findall('{\s*"stream"\s*:\s*"[^"]*"\s*}', line)
+            return get_build_id(lines, discard_logs)
+
+    return image_id, parsed_lines
+
+
+BuildCtxTuple = collections.namedtuple(
+    'BuildContext', ['format', 'path', 'dockerfile', 'job_params']
+)
+
+
+class BuildContext(BuildCtxTuple):
+    def __new__(cls, context_format,
+                path,
+                dockerfile='Dockerfile',
+                job_params=None):
+        ctx_tuple = super(BuildContext, cls)
+        return ctx_tuple.__new__(
+            cls,
+            context_format,
+            path,
+            dockerfile,
+            job_params,
+        )
+
+
+def make_context_from_tarball(path, dockerfile='Dockerfile'):
+    return BuildContext(
+        'tarball',
+        path,
+        dockerfile=dockerfile,
+        job_params={
+            'encoding': 'gzip',
+            'custom_context': True,
+            'fileobj': open(path)
+        }
+    )
+
+
+def make_context_from_dockerfile(path, dockerfile='Dockerfile'):
+    return BuildContext(
+        'dockerfile',
+        path=path,
+        dockerfile=dockerfile,
+        job_params={'fileobj': open(path, 'r')},
+    )
+
+
+def make_context_from_url(path, dockerfile='Dockerfile'):
+    return BuildContext(
+        'remote',
+        path,
+        dockerfile=dockerfile,
+        job_params={},
+    )
+
+
+def make_context_from_directory(path, dockerfile='Dockerfile'):
+    return BuildContext(
+        'directory',
+        path,
+        dockerfile=dockerfile,
+        job_params={}
+    )
```

Comment:
What do you think about making these factory classmethods on `BuildContext` instead of separate functions?

```python

class BuildContext(BuildCtxTuple):

    @classmethod
    def from_tarball(cls, path, dockerfile='Dockerfile'):
       return cls(...)

    @classmethod
    def from_dockerfile(cls, ...):
        ...
```